spec.html

```diff
@@ -801,6 +822,7 @@ <h1>Resizable ArrayBuffer and growable SharedArrayBuffer Guidelines</h1>
     <p>The following are guidelines for ECMAScript implementers implementing growable SharedArrayBuffer.</p>
     <p>We recommend growable SharedArrayBuffer be implemented as in-place growth via reserving virtual memory up front.</p>
     <p>Because grow operations can be concurrent with memory accesses on a growable SharedArrayBuffer, the constraints of the memory model require that even unordered accesses do not "tear" (bits of their values will not be mixed). In practice, this means the underlying data block of a growable SharedArrayBuffer cannot be grown by being copied without stopping the world. We do not recommend stopping the world as an implementation strategy because it introduces a serialization point and is slow.</p>
+    <p>Virtual memory pages should be zero-filled-on-demand so that new memory when growing are zeroed. If zero-fill-on-demand pages are not supported, then SharedArrayBuffer.prototype.grow must be implemented with a critical section that ensures new memory is zeroed before atomically updating the byte length. Any contended attempts to enter the critical section must throw.</p>
```

Comment:
There's a lot of subtlety around the critical section implementation. The implementation would probably need to additionally [`membarrier`](https://man7.org/linux/man-pages/man2/membarrier.2.html) all application threads after zeroing memory but before writing the new length (although I need to think about this more, so take with a big grain of salt for now).

There's also the option of really allocating all the zero-pages upfront.

Reply:
> There's a lot of subtlety around the critical section implementation. The implementation would probably need to additionally membarrier all application threads after zeroing memory but before writing the new length (although I need to think about this more, so take with a big grain of salt for now).

Hm, good point. Do you think it needs more elaboration here? In reality I imagine all engines depend on ZFOD pages outside of niche ones for specific contexts (like XS for embedded).

> There's also the option of really allocating all the zero-pages upfront.

Good point.

Reply:
My completely personal preference would be to be somewhat more abstract - e.g.

"Grown memory must appear zeroed from the moment of its creation, including to any racy concurrent accesses."

If some implementation explanation is still needed,

"This can be accomplished through zero-filled-on-demand virtual memory pages, or through careful synchronization if manually zeroing memory."